lib/gc/ExecutionEngine/CPURuntime/MemoryPool.cpp

```diff
@@ -0,0 +1,262 @@
+//===-- MemoryPool.cpp - Runtime allocator memory pool ----------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <cassert>
+#include <iostream>
+#include <memory.h>
+#include <stdexcept>
+#include <stdio.h>
+#include <stdlib.h>
+
+#ifdef _WIN32
+#include <Windows.h>
+#else
+#include <sys/mman.h>
+#include <unistd.h>
+#endif
+
+#ifdef _MSC_VER
+#define __builtin_expect(EXP_, C) (EXP_)
+#endif
+
+#define likely(x) __builtin_expect(!!(x), 1)
+#define unlikely(x) __builtin_expect(!!(x), 0)
+namespace {
+// 4MB
+constexpr size_t threadlocalChunkSize = 4 * 1024 * 1024;
+// 16MB
+constexpr size_t mainChunkSize = 16 * 1024 * 1024;
+
+static constexpr size_t defaultAlignment = 64;
+
+size_t getOsPageSize() {
+#ifdef _WIN32
+  // fix-me: (win32) impl
+  return 4096;
+#else
+  static size_t v = getpagesize();
+  return v;
+#endif
+}
+
+static constexpr size_t divideAndCeil(size_t x, size_t y) {
+  return (x + y - 1) / y;
+}
```

Comment:
Equivalent to `llvm::divideCeil()`?

Reply:
`llvm::divideCeil()` is equivalent to, which has an extra `/ Align * Align`
```
inline uint64_t alignTo(uint64_t Value, uint64_t Align) {
  assert(Align != 0u && "Align can't be 0.");
  return (Value + Align - 1) / Align * Align;
}

inline uint64_t divideCeil(uint64_t Numerator, uint64_t Denominator) {
  return alignTo(Numerator, Denominator) / Denominator;
}

```